Gems/Terrain/Code/Source/Components/TerrainWorldRendererComponent.cpp

```diff
@@ -0,0 +1,225 @@
+    {
+    }
+
+    TerrainWorldRendererComponent::~TerrainWorldRendererComponent()
+    {
+        if (m_terrainRendererActive)
+        {
+            Deactivate();
+        }
+    }
+
+    void TerrainWorldRendererComponent::Activate()
+    {
+        // On component activation, register the terrain feature processor with Atom and the default scene.
+
+        AZ::RPI::FeatureProcessorFactory::Get()->RegisterFeatureProcessor<Terrain::TerrainFeatureProcessor>();
+        if (auto rpi = AZ::RPI::RPISystemInterface::Get(); rpi)
+        {
+            if (auto defaultScene = rpi->GetDefaultScene(); defaultScene)
+            {
+                AZ::RPI::Scene* scene = defaultScene.get();
+                scene->EnableFeatureProcessor<Terrain::TerrainFeatureProcessor>();
+            }
+        }
+
+        AzFramework::Terrain::TerrainDataNotificationBus::Handler::BusConnect();
+        m_terrainRendererActive = true;
+    }
+
+    void TerrainWorldRendererComponent::Deactivate()
+    {
+        // On component deactivation, unregister the feature processor and remove it from the default scene.
+
+        m_terrainRendererActive = false;
+        AzFramework::Terrain::TerrainDataNotificationBus::Handler::BusDisconnect();
+
+        if (auto rpi = AZ::RPI::RPISystemInterface::Get(); rpi)
+        {
+            if (auto defaultScene = rpi->GetDefaultScene(); defaultScene)
+            {
+                AZ::RPI::Scene* scene = defaultScene.get();
+                if (scene->GetFeatureProcessor<Terrain::TerrainFeatureProcessor>())
+                {
+                    scene->DisableFeatureProcessor<Terrain::TerrainFeatureProcessor>();
+                }
+            }
+        }
+
+        AZ::RPI::FeatureProcessorFactory::Get()->UnregisterFeatureProcessor<Terrain::TerrainFeatureProcessor>();
+    }
+
+    bool TerrainWorldRendererComponent::ReadInConfig(const AZ::ComponentConfig* baseConfig)
+    {
+        if (auto config = azrtti_cast<const TerrainWorldRendererConfig*>(baseConfig))
+        {
+            m_configuration = *config;
+            return true;
+        }
+        return false;
+    }
+
+    bool TerrainWorldRendererComponent::WriteOutConfig(AZ::ComponentConfig* outBaseConfig) const
+    {
+        if (auto config = azrtti_cast<TerrainWorldRendererConfig*>(outBaseConfig))
+        {
+            *config = m_configuration;
+            return true;
+        }
+        return false;
+    }
+
+    void TerrainWorldRendererComponent::OnTerrainDataDestroyBegin()
+    {
+        // If the terrain is being destroyed, remove all existing terrain data from the feature processor.
+
+        if (auto rpi = AZ::RPI::RPISystemInterface::Get(); rpi)
+        {
+            if (auto defaultScene = rpi->GetDefaultScene(); defaultScene)
+            {
+                const AZ::RPI::Scene* scene = defaultScene.get();
+                if (auto terrainFeatureProcessor = scene->GetFeatureProcessor<TerrainFeatureProcessor>(); terrainFeatureProcessor)
```

Comment:
There's several calls in this file that re-lookup the terrain feature processor even though this component creates it. Why not just cache a reference to it? If you're worried about it being deleted from under you, you could at least cache the correct AZFramework::Scene so you have a more direct way to look it up. Same as above, `GetDefaultScene()`isn't a good idea here.

Reply:
Fixed!  The deletion concern came from the create/destroy previously living in a separate file from the usage.  Now that they're together, the caching is trivial and safe.